pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/PartitionDedupMetadataManager.java

```diff
@@ -0,0 +1,154 @@
+    while (recordInfoIterator.hasNext()) {
+      RecordInfo recordInfo = recordInfoIterator.next();
+      _primaryKeySet.put(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), segment);
+    }
+    _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.DEDUP_PRIMARY_KEYS_COUNT,
+        _primaryKeySet.size());
+  }
+
+  public void removeSegment(IndexSegment segment) {
+    // TODO(saurabh): Explain reload scenario here
+    Iterator<RecordInfo> recordInfoIterator = getRecordInfoIterator(segment, _primaryKeyColumns);
+    while (recordInfoIterator.hasNext()) {
+      RecordInfo recordInfo = recordInfoIterator.next();
+      _primaryKeySet.compute(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction),
+          (primaryKey, currentSegment) -> {
+            if (currentSegment == segment) {
+              return null;
+            } else {
+              return currentSegment;
+            }
+          });
+    }
+    _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.DEDUP_PRIMARY_KEYS_COUNT,
+        _primaryKeySet.size());
+  }
+
+  @VisibleForTesting
+  public static Iterator<RecordInfo> getRecordInfoIterator(IndexSegment segment, List<String> primaryKeyColumns) {
+    Map<String, PinotSegmentColumnReader> columnToReaderMap = new HashMap<>();
+    for (String primaryKeyColumn : primaryKeyColumns) {
+      columnToReaderMap.put(primaryKeyColumn, new PinotSegmentColumnReader(segment, primaryKeyColumn));
+    }
+    int numTotalDocs = segment.getSegmentMetadata().getTotalDocs();
+    int numPrimaryKeyColumns = primaryKeyColumns.size();
+    return new Iterator<RecordInfo>() {
+      private int _docId = 0;
+
+      @Override
+      public boolean hasNext() {
+        return _docId < numTotalDocs;
+      }
+
+      @Override
+      public RecordInfo next() {
+        Object[] values = new Object[numPrimaryKeyColumns];
+        for (int i = 0; i < numPrimaryKeyColumns; i++) {
+          Object value = columnToReaderMap.get(primaryKeyColumns.get(i)).getValue(_docId);
+          if (value instanceof byte[]) {
+            value = new ByteArray((byte[]) value);
+          }
+          values[i] = value;
+        }
+        PrimaryKey primaryKey = new PrimaryKey(values);
+        return new RecordInfo(primaryKey, _docId++, null);
+      }
+    };
+  }
+
+  private synchronized void waitTillAllSegmentsLoaded() {
+    while (!TableState.isAllSegmentsLoaded(_helixManager, _tableNameWithType)) {
+      LOGGER.info("Sleeping 1 second waiting for all segments loaded for partial-upsert table: {}", _tableNameWithType);
+      try {
+        //noinspection BusyWait
+        Thread.sleep(1000L);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    _allSegmentsLoaded = true;
+  }
+
+  public boolean checkRecordPresentOrUpdate(RecordInfo recordInfo, IndexSegment indexSegment) {
+    if (!_allSegmentsLoaded) {
```

Comment:
Can you help me understand the thread safety concerns with this? I don't any, single threaded or multi threaded.

Infact, moving this if check inside `waitTillAllSegmentsLoaded()` would lead to unnecessary serialization even when all segments have already been loaded. Even in single threaded env, that's a heavy lock acquisition cost, when `_allSegmentsLoaded` is already true.

To the point where, I think we should reduce the critical section here https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java#L74, once `_allSegmentsLoaded` has been set to true, no need to enter a syncronized block.

Do let me know your thoughts

Reply:
I misused the word thread-safety. I was suggesting adding an extra if check in the synchronized block to avoid potential unnecessary checks when multiple threads invoke `waitTillAllSegmentsLoaded()`.
Good point on reducing the critical section in `PartialUpsertHandler`. We should first check the flag, then enter the critical section